### PR TITLE
[WIP] Fluxible Router

### PR DESCRIPTION
**Pull request by mridgway**
Background discussion in #41

This is an attempt at presenting a more Flux-like router. There is no longer need for a custom plugin, we simply use a `RouteStore` as the source of truth and the functionality we need. This actually simplifies the setup process as well and allows our `NavLink` component to know whether it is active or not.

 * No more plugin or childContext; logic and state have both moved to the `RouteStore`
   * Developer could entirely replace the `RouteStore` as long as it adheres to the same interface:
      * Handles the `NAVIGATE_START` event
      * Exposes a `getCurrentRoute()` method that returns a route object (to be defined)
      * Exposes a `makePath(name, params)` method
      * Exposes an `isActive(url)` method
 * `NavLink` knows whether it is currently active and has properties to support changing styles or class.
    * `activeStyles`: adds the styles hash if the route is active
    * `activeClass`: adds the class if the route is active
 * Use immutable for the route state in the `RouteStore`. This should be fixed up a little, but I think there are benefits to doing this. It might even make sense to require the route config that's passed in to be immutable.
 * This would support both static and dynamic routes through a flux dispatch similarly to @localnerve's https://github.com/yahoo/fluxible-plugin-routr/pull/14 (sorry for possibly replacing your library; you inspired me).
   * Dynamic routes are built-in. If you dispatch a `RECEIVE_ROUTES` event with the route config, it will be used for the next `NAVIGATE_START`
   * Static routes are provided by a helper `RouteStore.withStaticRoutes(routeConfig)` that creates a `StaticRouteStore` that extends `RouteStore`.

You can see the fluxible router example changes at https://github.com/yahoo/flux-examples/compare/fluxible-router

Things to do: 
 * [x] ~~Fix `RouterMixin` not having access to the `navigate` key~~
 * [x] Incorporate PRs that have been merged into both flux-router-component and fluxible-plugin-routr since I started this work
 * [x] Add more unit tests to cover newly added APIs like `RouteStore`
 * [x] Add `handleHistory` higher-order component for handling history state
 * [x] Add `handleRoute` higher-order component for listening to route changes
 * [x] Full API documentation
    * [x] `handleHistory`
    * [x] `handleRoute`
    * [x] `navigateAction`
    * [x] `RouteStore`
 * [x] Use updated `connectToStores` interface
 * [x] Rename error `status` property to `statusCode` in `navigateAction`
 * [x] Rebase to integrate latest changes (#94 and #91)
 * [ ] Rename `flux-router-component` to `fluxible-router`
 * [ ] Deprecate `fluxible-plugin-routr` on ynpm (add message to README)

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

**Comment by jedireza** [comment]
:clap: this is looking great

**Comment by localnerve** [comment]
:blush: This is Fluxibly Delicious :tm:

File: README.md [hunk 1]
```diff
@@ -1,142 +1,115 @@
...
-                {links}
-            </ul>
-        );
-
-    }
+// app.js
+var FluxibleApp = require('fluxible');
+var RouteStore = require('fluxible-router').RouteStore;
+var app = new FluxibleApp({
+    appComponent: require('./components/App.jsx')
```
Comment (lingyan) [inline README.md]: Use `component`

File: README.md [hunk 2]
```diff
@@ -1,142 +1,115 @@
...
+// components/App.jsx
+var FluxibleMixin = require('fluxible').FluxibleMixin;
+var RouterMixin = require('fluxible-router').RouterMixin;
+var NavLink = require('fluxible-router').NavLink;
+module.exports = React.createClass({
+    // The RouterMixin should be mixed in to your top level component
+    // The router mixin handles changing the window pushState when a navigate happens
+    mixins: [FluxibleMixin, RouterMixin],
+    render: function () {
+        // Get the handler form the current route
```
Comment (lingyan) [inline README.md]: from

File: lib/RouteStore.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    _matchRoute: function (url, options) {
+        var self = this;
+        var route = self.getRouter().getRoute(url, options);
+        if (!route) {
+            return null;
+        }
+
+        var matchedUrl = route.url;
+        route = Immutable.fromJS(route.config).withMutations(function (r) {
+            r.set('params', Immutable.fromJS(route.params));
```
Comment (lingyan) [inline lib/RouteStore.js]: name anonymous function

File: lib/RouteStore.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    _parseQueryString: function (url) {
+        var matches = url.match(searchPattern);
+        var search;
+        if (matches) {
+            search = matches[1];
+        }
+        return (search && queryString.parse(search)) || {};
+    },
+    handleReceiveRoutes: function (payload) {
+        this.routes = objectAssign(this.routes || {}, payload);
```
Comment (lingyan) [inline lib/RouteStore.js]: also make routes immutable?

File: lib/RouteStore.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
+var createStore = require('fluxible/utils/createStore');
+var Router = require('routr');
+var queryString = require('query-string');
+var searchPattern = /\?([^\#]*)/;
+var util = require('util');
```
Comment (lingyan) [inline lib/RouteStore.js]: use https://github.com/isaacs/inherits instead?

File: lib/NavLink.js [hunk 6]
```diff
@@ -6,20 +6,68 @@
...
+    },
+    componentWillReceiveProps: function (nextProps) {
+        var href = this._getHrefFromProps(nextProps);
+        this.setState({
+            href: this._getHrefFromProps(nextProps),
+            isActive: this._getRouteStore().isActive(href)
+        });
+    },
+    componentDidMount: function () {
+        this._getRouteStore().addChangeListener(this._onStoreChange);
```
Comment (lingyan) [inline lib/NavLink.js]: One thing to call out, there could be more than 10 NavLink's on the page, which will trigger the max listener count warning with event emitter.

File: lib/RouteStore.js [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+var objectAssign = require('object-assign');
+var Immutable = require('immutable');
+
+var RouteStore = createStore({
+    storeName: 'RouteStore',
+    handlers: {
+        'NAVIGATE_START': 'handleNavigateStart',
```
Comment (lingyan) [inline lib/RouteStore.js]: we should think about how RouteStore reacts to `NAVIGATE_SUCCESS` and `NAVIGATE_FAILURE` events.

File: tests/unit/lib/navigateAction.js [hunk 8]
```diff
@@ -0,0 +1,160 @@
+/**
+ * Copyright 2014, Yahoo! Inc.
```
Comment (lingyan) [inline tests/unit/lib/navigateAction.js]: fix copyright headers to have 2014-2015, and remove `!`

File: lib/RouteStore.js [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        'RECEIVE_ROUTES': 'handleReceiveRoutes'
```
Comment (lingyan) [inline lib/RouteStore.js]: Document this `RECEIVE_ROUTES` event.

File: README.md [hunk 10]
```diff
@@ -1,142 +1,115 @@
...
+        handler: require('../components/About.jsx')
+    },
+    user: {
+        method: 'GET',
+        path: '/user/:id',
+        handler: require('../components/User.jsx')
     }
-});
+};
 ```
```
Comment (lingyan) [inline README.md]: add a section or example about dynamic route store

File: README.md [hunk 11]
```diff
@@ -1,142 +1,115 @@
...
+        path: '/about',
+        handler: require('../components/About.jsx')
+    },
+    user: {
+        method: 'GET',
+        path: '/user/:id',
+        handler: require('../components/User.jsx')
     }
-});
+};
```
Comment (lingyan) [inline README.md]: add a section with example about dynamic route store

**Comment by lingyan** [comment]
I definitely like the direction of where this is going.  Using RouteStore definitely brings in more fluxibility and flexibility ;)

File: lib/RouteStore.js (hunk 4 above)
Comment (mridgway) [inline lib/RouteStore.js]: I found that this actually just made things more complicated and added unnecessary overhead. I could only see this being useful if routr handled immutable route config or if one would expect the routes to change a lot throughout the store's lifetime.

File: lib/NavLink.js (hunk 6 above)
Comment (mridgway) [inline lib/NavLink.js]: I can set the maxEventListeners for this specific instance higher.

File: lib/RouteStore.js [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        this.currentRoute = this._matchRoute(this.currentUrl, {method: 'GET'});
+        this.currentNavigate = state.currentNavigate;
+    }
+});
+
+RouteStore.withStaticRoutes = function (staticRoutes) {
+    function StaticRouteStore() {
+        RouteStore.apply(this, arguments);
+    }
+    util.inherits(StaticRouteStore, RouteStore);
```
Comment (lingyan) [inline lib/RouteStore.js]: use `inherits`

File: README.md (hunk 2 above)
Comment (lingyan) [inline README.md]: Add doc about RouterComponent

**Comment by lance-anderson** [comment]
:+1: +1 for adding support for handlers:
```handler: require('../components/About.jsx')```

Does this PR also include support for dynamic routes so that one doesn't have to manually configure the above handler?

e.g.
a request to ```/about/jobs``` auto-magically routes to ```../components/about/Jobs.jsx```

or something along those lines?

cause if it does that's +100 :smile:

**Comment by mridgway** [comment]
@lance-anderson No auto-magic support like that for now.

**Comment by mridgway** [comment]
I've pushed more changes including a couple of higher-order components: `handleHistory` and `handleRoute`.

`handleHistory` is a wrapper around your app component for managing the `window.location` and history state.

`handleRoute` is a wrapper that can be used to make any component aware of the `currentRoute`. It is internally used by `handleHistory` and `NavLink`, but could be useful for other purposes and is therefore exposed in the main exports.

[flux-example PR](https://github.com/yahoo/flux-examples/pull/119) has been updated with latest changes from this PR and Fluxible@0.4.x.

I just need to finish up documentation at this point.

File: lib/navigateAction.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        return;
+    }
+    debug('executing', payload);
+
+    var route = routeStore.getCurrentRoute();
+
+    if (!route) {
+        context.dispatch('NAVIGATE_FAILURE', payload);
+        var err = new Error('Url does not exist');
+        err.status = 404;
```
Comment (mridgway) [inline lib/navigateAction.js]: `err.statusCode = 404`

Comment (lingyan) [inline lib/navigateAction.js]: Yes. Lets make the change and document it in upgrade notes.

**Comment by jedireza** [comment]
Ok I rebased from master and fixed all the conflicts. Here is the diff of my current `fluxible-router` branch and this one:

https://github.com/yahoo/flux-router-component/compare/fluxible-router...jedireza:fluxible-router

**Comment by mridgway** [comment]
This is being released under a new repo available at https://github.com/yahoo/fluxible-router

**Pull request closed without merging**